### PR TITLE
emacsPackages.p4-16-mode: fix build

### LINKED ISSUE #301795: Build failure: Emacs with ELPA package `p4-16-mode`
### Steps To Reproduce

With `flake.nix`:

```nix
{
  description = "A very basic flake";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs?ref=nixos-unstable";
  };

  outputs = { self, nixpkgs }:
    let
      system = "x86_64-linux";
      pkgs = import nixpkgs { inherit system; };
    in
    {
      devShells."${system}".default = with pkgs; mkShell {
        packages = [
          ((emacsPackagesFor emacs-gtk).emacsWithPackages (e: [ e.p4-16-mode ]))
        ];
      };
    };
}
```

### Build log

```
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
Running phase: buildPhase
no Makefile or custom buildPhase, doing nothing
@nix { "action": "setPhase", "phase": "installPhase" }
Running phase: installPhase
Parsing tar file... 
Parsing tar file...done

Error: wrong-type-argument (arrayp nil)
  mapbacktrace(#f(compiled-function (evald func args flags) #<bytecode -0x51fa72ed2510e41>))
  debug-early-backtrace()
  debug-early(error (wrong-type-argument arrayp nil))
  tar--describe-as-link(nil)
  tar--check-descriptor(nil)
  tar-get-file-descriptor("p4-16-mode-0.3/p4-pkg.el")
  package-tar-file-info()
  (if (derived-mode-p 'tar-mode) (package-tar-file-info) (package-buffer-info))
  (let ((pkg-desc (if (derived-mode-p 'tar-mode) (package-tar-file-info) (package-buffer-info)))) (p>
  elpa2nix-install-from-buffer()
  (progn (if is-tar (insert-file-contents-literally file) (insert-file-contents file)) (if is-tar (p>
  (unwind-protect (progn (if is-tar (insert-file-contents-literally file) (insert-file-contents file>
  (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (if is-tar (insert-file-conte>
  (let ((temp-buffer (generate-new-buffer " *temp*" t))) (save-current-buffer (set-buffer temp-buffe>
  (let ((is-tar (string-match "\\.tar\\'" file))) (let ((temp-buffer (generate-new-buffer " *temp*" >
  elpa2nix-install-file("/nix/store/m5gj443yzg81163fncw1rb58c81kz8rl-p4-16-mode-0.3.tar")
  (progn (setq package-user-dir elpa) (elpa2nix-install-file archive))
  (let ((archive x0) (elpa x2)) (progn (setq package-user-dir elpa) (elpa2nix-install-file archive)))
  (if (null x3) (let ((archive x0) (elpa x2)) (progn (setq package-user-dir elpa) (elpa2nix-install->
  (let* ((x2 (car-safe x1)) (x3 (cdr-safe x1))) (if (null x3) (let ((archive x0) (elpa x2)) (progn (>
  (if (consp x1) (let* ((x2 (car-safe x1)) (x3 (cdr-safe x1))) (if (null x3) (let ((archive x0) (elp>
  (let* ((x0 (car-safe command-line-args-left)) (x1 (cdr-safe command-line-args-left))) (if (consp x>
  (if (consp command-line-args-left) (let* ((x0 (car-safe command-line-args-left)) (x1 (cdr-safe com>
  elpa2nix-install-package()
  command-line-1(("-l" "/nix/store/4jj63z4v1xp13rh2md053dccq920hd45-elpa2nix.el" "-f" "elpa2nix-inst>
  command-line()
  normal-top-level()
Wrong type argument: arrayp, nil
```

### Additional context

Directly installing this package from within Emacs (using `package-install`) does not cause this error.

The execution leading to this build error is:
1. `elpa2nix-install-from-buffer` in `elpa2nix.el` calls `(package-tar-file-info)`
2. ... which calls `package--description-file` on the unpacked directory name (`p4-16-mode-0.3`)
3. ... which parses this name to get the "package name" and append `-pkg.el` to get the "package description file name"
4. but wrongly thinks the "package name" is "p4" instead of "p4-16-mode", and looks for `p4-pkg.el` (instead of `p4-16-mode-pkg.el`), which does not exist.

### Notify maintainers

@AndersonTorres @jian-lin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by jian-lin**
https://hydra.nixos.org/build/269588786

Closes https://github.com/NixOS/nixpkgs/issues/301795

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/editors/emacs/elisp-packages/nongnu-devel-packages.nix [hunk 1]
```diff
@@ -35,7 +35,23 @@ let
 
       super = imported;
 
-      overrides = { };
+      overrides = {
```
Comment (adisbladis) [inline pkgs/applications/editors/emacs/elisp-packages/nongnu-devel-packages.nix:38]: Quite complex and tricky..
Future note: Should we dedup overrides between nongnu stable/devel?

Comment (jian-lin) [reply pkgs/applications/editors/emacs/elisp-packages/nongnu-devel-packages.nix:38]: Hmm, the complexity is due to `src` being an unpacked tarball.  I would say the core of this workaround is simple:

```console
cp --verbose $content_directory/p4-16-mode-pkg.el $content_directory/p4-pkg.el
```

> Should we dedup overrides between nongnu stable/devel?

I think we should and it is tracked in https://github.com/NixOS/nixpkgs/issues/269550.

**Comment by hexclover** [comment]
It seems this fix hasn't been applied to the [Emacs overlay](https://github.com/nix-community/emacs-overlay). Can this be ported there as well?

**Comment by jian-lin** [comment]
Emacs overlay is not relevant here.

What matters is that whether your Nixpkgs contains this PR or not. Note that this PR has not been backported to the stable release (`nixos-24.05`) yet.  Do you use the stable release?

**Pull request merged**